rdr_service/services/data_quality.py

```diff
@@ -120,19 +127,39 @@ def run_data_quality_checks(self, for_data_since: datetime = None):
                 Participant.signUpTime,
                 Participant.suspensionTime,
                 Participant.withdrawalAuthored,
-                func.count(QuestionnaireResponseAnswer.questionnaireResponseAnswerId),
-                QuestionnaireHistory.created
+                func.min(all_matching_questionnaire.created),
+                func.group_concat(Code.value.distinct())
             )
             .join(Participant)
-            .join(QuestionnaireHistory)
-            .outerjoin(QuestionnaireResponseAnswer)
+            .outerjoin(
+                QuestionnaireConcept,
+                and_(
+                    QuestionnaireConcept.questionnaireId == QuestionnaireResponse.questionnaireId,
+                    QuestionnaireConcept.questionnaireVersion == QuestionnaireResponse.questionnaireVersion
+                )
+            )
+            .outerjoin(
+                Code,
+                Code.codeId == QuestionnaireConcept.codeId
+            )
+            .outerjoin(
+                all_matching_concepts,
+                all_matching_concepts.codeId == Code.codeId
+            )
+            .outerjoin(
+                all_matching_questionnaire,
+                and_(
+                    all_matching_questionnaire.questionnaireId == all_matching_concepts.questionnaireId,
+                    all_matching_questionnaire.version == all_matching_concepts.questionnaireVersion
+                )
+            )
             .group_by(QuestionnaireResponse.questionnaireResponseId)
         )
         if for_data_since is not None:
             query = query.filter(QuestionnaireResponse.created >= for_data_since)
 
         for response_id, created_time, authored_time, participant_signup_time, suspension_datetime,\
-                withdrawal_datetime, answer_count, questionnaire_created_time in query.all():
+                withdrawal_datetime, min_questionnaire_created_time, module_code in query.all():
```

Comment:
Not in this PR, but should we write these logs to a table for record-keeping and analyses in the future?

Reply:
I think that's a good idea, but maybe we can wait to see how often errors come up and how they are handled so we'd know how much value there would be in it.